source/sundials/kinsol.cc

```diff
@@ -0,0 +1,336 @@
+      int err = 0;
+      if (solver.residual)
+        err = solver.residual(*src_yy, *dst_FF);
+      else if (solver.iteration_function)
+        err = solver.iteration_function(*src_yy, *dst_FF);
+      else
+        Assert(false, ExcInternalError());
+
+      copy(FF, *dst_FF);
+
+      return err;
+    }
+
+
+
+    template<typename VectorType>
+    int t_kinsol_setup_jacobian(KINMem kinsol_mem)
+    {
+      KINSOL<VectorType> &solver = *static_cast<KINSOL<VectorType> *>(kinsol_mem->kin_user_data);
+      GrowingVectorMemory<VectorType> mem;
+
+      typename VectorMemory<VectorType>::Pointer src_ycur(mem);
+      solver.reinit_vector(*src_ycur);
+
+      typename VectorMemory<VectorType>::Pointer src_fcur(mem);
+      solver.reinit_vector(*src_fcur);
+
+      copy(*src_ycur, kinsol_mem->kin_uu);
+      copy(*src_fcur, kinsol_mem->kin_fval);
+
+      int err = solver.setup_jacobian(*src_ycur, *src_fcur);
+      return err;
+    }
+
+
+
+    template<typename VectorType>
+    int t_kinsol_solve_jacobian(KINMem kinsol_mem,
+                                N_Vector x,
+                                N_Vector b,
+                                realtype *sJpnorm,
+                                realtype *sFdotJp)
+    {
+      KINSOL<VectorType> &solver = *static_cast<KINSOL<VectorType> *>(kinsol_mem->kin_user_data);
+      GrowingVectorMemory<VectorType> mem;
+
+      typename VectorMemory<VectorType>::Pointer src_ycur(mem);
+      solver.reinit_vector(*src_ycur);
+
+      typename VectorMemory<VectorType>::Pointer src_fcur(mem);
+      solver.reinit_vector(*src_fcur);
+
+      copy(*src_ycur, kinsol_mem->kin_uu);
+      copy(*src_fcur, kinsol_mem->kin_fval);
+
+      typename VectorMemory<VectorType>::Pointer src(mem);
+      solver.reinit_vector(*src);
+
+      typename VectorMemory<VectorType>::Pointer dst(mem);
+      solver.reinit_vector(*dst);
+
+      copy(*src, b);
+
+      int err = solver.solve_jacobian_system(*src_ycur, *src_fcur,
+                                             *src,*dst);
+      copy(x, *dst);
+
+      *sJpnorm = N_VWL2Norm(b, kinsol_mem->kin_fscale);
+      N_VProd(b, kinsol_mem->kin_fscale, b);
+      N_VProd(b, kinsol_mem->kin_fscale, b);
+      *sFdotJp = N_VDotProd(kinsol_mem->kin_fval, b);
+
+      return err;
+    }
+  }
+
+  template <typename VectorType>
+  KINSOL<VectorType>::KINSOL(const AdditionalData &data, const MPI_Comm mpi_comm) :
+    data(data),
+    kinsol_mem(nullptr),
+    communicator(Utilities::MPI::duplicate_communicator(mpi_comm))
+  {
+    set_functions_to_trigger_an_assert();
+  }
+
+
+
+  template <typename VectorType>
+  KINSOL<VectorType>::~KINSOL()
+  {
+    if (kinsol_mem)
+      KINFree(&kinsol_mem);
+    MPI_Comm_free(&communicator);
+  }
+
+
+
+  template <typename VectorType>
+  unsigned int KINSOL<VectorType>::solve(VectorType &initial_guess_and_solution)
+  {
+    unsigned int system_size = initial_guess_and_solution.size();
+    unsigned int local_system_size = system_size;
```

Comment:
I would move the declaration of `local_system_size` inside the block guarded with `#ifdef DEAL_II_WITH_MPI` otherwise you get a warning of unused variable